### PR TITLE
Feature/improve halem

### PR DESCRIPTION
**Pull request by Pietervanhalem**

### CODE REVIEW

File: openclsim/core.py [hunk 1]
```diff
@@ -1194,6 +1200,61 @@ def __init__(
...
+            # Duration of sailing full
+            _, TT, _ = self.optimization_func(
+                start, stop, t0, self.env.Roadmap.vship[i, -1], self.env.Roadmap
+            )
+
+            # Duration of sailing empty + loading + sailing full + unloading
+            duration_unloading = self.unloading(
+                None, None, self.container.capacity * self.loadfactors[i]
+            )
+            TTT += (TT[-1] - TT[0]) + duration_unloading
```
Comment (uijl) [inline openclsim/core.py]: This doesn't seem necessary for further computations.

File: openclsim/model.py [hunk 2]
```diff
@@ -302,6 +314,16 @@ def single_run_process(
             amount, mover.check_optimal_filling(loader, unloader, origin, destination)
         )
 
+    if (
+        isinstance(mover, core.Routeable)
+        and mover.optimize_route == True
+        and type(mover.loadfactors) != type(None)
+    ):
+        opt_LF = mover.check_optimal_filling_Roadmap(
```
Comment (uijl) [inline openclsim/model.py]: maybe add a comment or increase the length of variable opt_LF so it is clear what it returns

File: setup.py [hunk 3]
```diff
@@ -37,7 +37,7 @@
     "Flask-cors",
     "sphinx_rtd_theme",
     "Dill",
-    "halem==0.2.0",
+    "halem==0.3.0",
```
Comment (uijl) [inline setup.py]: does this break the openclsim notebooks from working with azure?

File: simulations/cdee1a97df0e954187f50a5615a4ee3c_equipment.csv [hunk 4]
```diff
@@ -0,0 +1,2 @@
+EquipmentID,EquipmentName
```
Comment (uijl) [inline simulations/cdee1a97df0e954187f50a5615a4ee3c_equipment.csv]: Please delete this, and all other *.csv test results

File: tests/test_Halem.py [hunk 5]
```diff
@@ -507,3 +508,295 @@ def compute_v_provider(v_empty, v_full):
...
+    )  # We stop once there is nothing more to move
+    my_env.FG = FG
+    my_env.Roadmap = Roadmap
+    my_env.run()
+
+    opt_LF = len(
+        list(set(np.array(hopper.log["Value"])[np.array(hopper.log["Value"]) > 0]))
+    )
+
+    assert opt_LF < no_opt_LF
```
Comment (uijl) [inline tests/test_Halem.py]: Add some comments to explain what your test is about

File: openclsim/core.py (hunk 1 above)
Comment (Pietervanhalem) [reply openclsim/core.py]: This part is needed for the calculation of the optimal load factor

File: setup.py (hunk 3 above)
Comment (Pietervanhalem) [reply setup.py]: Yes it does, I will fix this and create a new pull request in the notebook repo

File: openclsim/model.py (hunk 2 above)
Comment (Pietervanhalem) [reply openclsim/model.py]: done

File: simulations/cdee1a97df0e954187f50a5615a4ee3c_equipment.csv (hunk 4 above)
Comment (Pietervanhalem) [reply simulations/cdee1a97df0e954187f50a5615a4ee3c_equipment.csv]: Done in the new commit

File: tests/test_Halem.py (hunk 5 above)
Comment (Pietervanhalem) [reply tests/test_Halem.py]: Done for all the HALEM tests in commit "implement change requests for pull request"

**Pull request merged**